### DISCUSSION TITLE
Plot voxelised point cloud by coordinate centers and xyz-span

### DISCUSSION THREAD

**Discussion by colabfrr** [pyvista/pyvista#4653] [Q&A] [answered]
Hello, 

I have already computed the centers of the voxels of a point cloud (the voxels have all the same dimensions: 0.2x0.2x0.2). Is there a way to visualize them as cubes in pyvista? So far i add (one by one) cubes with the centers and the {x,y,z}_length parameters to a plotter, is this the recommended way? Maybe is my PC specs, but for around 500 cubes, it gets a bit slow. 

Thanks,

**Comment by banesullivan** [comment]
Can you share your data and the code you have as a starting point? Or some other minimal example?

**Reply by colabfrr** [reply]
Hello, thank you for you answer.

Basically I am using a point cloud and then use the spconv function `generate_voxel_with_id` (https://github.com/traveller59/spconv/blob/125a194d895b1bc3ad6ff907bc72641548397b32/spconv/pytorch/utils.py#L91C9-L91C31), which for an initial point cloud of, say 26312 points, returns in my case 4 variables (the explanations of each is my understanding of them):

voxels: [6182, 200, 3] : these are the coordinates of the points grouped by voxel (each voxel holds up to 200 points)
indices: [6182, 3] : these are the coordinates of the centers of the voxels
num_point_per_voxel: [6182] : this holds the amount of points of each voxel
pc_voxel_ids: [26312] : these are the indices of the voxel that each point belongs to (-1 if does not belong to any voxel)

What i would like is a function that plots the centers as cubes, like `Plotter.add_points(points, style='points', **kwargs)` in which the style is "cubes" and i could enter the dimensions of the cubes too (similar to the second image of this example http://www.open3d.org/docs/latest/tutorial/Advanced/voxelization.html#From-point-cloud)

**Reply by MatthewFlamm** [reply, marked as answer]
It sounds like you might be doing something like this?

```py
pl = pv.Plotter()
for center, size in zip(centers, sizes):
    pl.add_mesh(pv.Cube(center, size, size, size))
```

If so, you can try putting them all into a MultiBlock first:

```py
pl = pv.Plotter()
mb_mesh = pv.MultiBlock([pv.Cube(center, size, size, size) for center, size in zip(centers, sizes)])
pl.add_composite(mb_mesh)
```

Another option is using glyph filter using a Cube geometry which may be the fastest since it all happens in c++. https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.glyph.html

**Reply by colabfrr** [reply]
Thank you, the MultiBlock solution works well for my case.
